src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobReader.cs

```diff
@@ -495,6 +531,90 @@ public Handle ReadTypeHandle()
         }
 
         private static readonly uint[] s_corEncodeTokenArray = new uint[] { TokenTypeIds.TypeDef, TokenTypeIds.TypeRef, TokenTypeIds.TypeSpec, 0 };
+
+        /// <summary>
+        /// Reads a constant value (see ECMA-335 Partition II section 22.9) from the current position.
+        /// </summary>
+        /// <exception cref="BadImageFormatException">Error while reading from the blob.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="typeCode"/> is not a valid <see cref="ConstantTypeCode"/>.</exception>
+        /// <returns>
+        /// Boxed constant value. To avoid allocating the object use Read* methods directly.
+        /// Constants of type <see cref="ConstantTypeCode.String"/> are encoded as UTF16 strings, use <see cref="ReadUTF16(int)"/> to read them.
+        /// </returns>
+        public object ReadConstant(ConstantTypeCode typeCode)
+        {
+            // Partition II section 22.9:
+            //
+            // Type shall be exactly one of: ELEMENT_TYPE_BOOLEAN, ELEMENT_TYPE_CHAR, ELEMENT_TYPE_I1, 
+            // ELEMENT_TYPE_U1, ELEMENT_TYPE_I2, ELEMENT_TYPE_U2, ELEMENT_TYPE_I4, ELEMENT_TYPE_U4, 
+            // ELEMENT_TYPE_I8, ELEMENT_TYPE_U8, ELEMENT_TYPE_R4, ELEMENT_TYPE_R8, or ELEMENT_TYPE_STRING; 
+            // or ELEMENT_TYPE_CLASS with a Value of zero  (23.1.16)
+
+            // In addition reads Decimal and DateTime types used in debug metadata.
+
+            switch (typeCode)
+            {
+                case ConstantTypeCode.Boolean:
+                    return ReadBoolean();
+
+                case ConstantTypeCode.Char:
+                    return ReadChar();
+
+                case ConstantTypeCode.SByte:
+                    return ReadSByte();
+
+                case ConstantTypeCode.Int16:
+                    return ReadInt16();
+
+                case ConstantTypeCode.Int32:
+                    return ReadInt32();
+
+                case ConstantTypeCode.Int64:
+                    return ReadInt64();
+
+                case ConstantTypeCode.Byte:
+                    return ReadByte();
+
+                case ConstantTypeCode.UInt16:
+                    return ReadUInt16();
+
+                case ConstantTypeCode.UInt32:
+                    return ReadUInt32();
+
+                case ConstantTypeCode.UInt64:
+                    return ReadUInt64();
+
+                case ConstantTypeCode.Single:
+                    return ReadSingle();
+
+                case ConstantTypeCode.Double:
+                    return ReadDouble();
+
+                case ConstantTypeCode.String:
+                    return ReadUTF16(Length);
```

Comment:
This should rather be RemainingBytes, usually the same as Length since the method is used on whole blobs but just in case it's not...